### PR TITLE
Initial support for async in routing driver

### PR DESCRIPTION
**Pull request by lutovich**

### CODE REVIEW

File: driver/src/main/java/org/neo4j/driver/internal/cluster/loadbalancing/LoadBalancingStrategy.java [hunk 1]
```diff
@@ -33,11 +33,15 @@
      */
     BoltServerAddress selectReader( BoltServerAddress[] knownReaders );
 
+    BoltServerAddress selectReaderAsync( BoltServerAddress[] knownReaders );
```
Comment (ali-ince) [inline driver/src/main/java/org/neo4j/driver/internal/cluster/loadbalancing/LoadBalancingStrategy.java:36]: and here

File: driver/src/main/java/org/neo4j/driver/internal/cluster/loadbalancing/LoadBalancer.java [hunk 2]
```diff
@@ -193,6 +286,21 @@ private BoltServerAddress selectAddress( AccessMode mode, AddressSet servers )
         }
     }
 
+    private BoltServerAddress selectAddressAsync( AccessMode mode, AddressSet servers )
```
Comment (ali-ince) [inline driver/src/main/java/org/neo4j/driver/internal/cluster/loadbalancing/LoadBalancer.java:289]: Do we need this sync/async distinction at this level? Do you have plans to implement selection differently in the future?

File: driver/src/main/java/org/neo4j/driver/internal/cluster/loadbalancing/RoundRobinLoadBalancingStrategy.java [hunk 3]
```diff
@@ -47,11 +47,23 @@ public BoltServerAddress selectReader( BoltServerAddress[] knownReaders )
     }
 
     @Override
+    public BoltServerAddress selectReaderAsync( BoltServerAddress[] knownReaders )
```
Comment (ali-ince) [inline driver/src/main/java/org/neo4j/driver/internal/cluster/loadbalancing/RoundRobinLoadBalancingStrategy.java:50]: and here

File: driver/src/main/java/org/neo4j/driver/internal/cluster/loadbalancing/LoadBalancer.java [hunk 4]
```diff
@@ -44,56 +48,83 @@
...
     {
         this.connections = connections;
+        this.asyncConnectionPool = asyncConnectionPool;
         this.routingTable = routingTable;
         this.rediscovery = rediscovery;
         this.loadBalancingStrategy = loadBalancingStrategy;
         this.log = log;
 
-        refreshRoutingTable();
+        if ( connections != null )
```
Comment (ali-ince) [inline driver/src/main/java/org/neo4j/driver/internal/cluster/loadbalancing/LoadBalancer.java:95]: does this depend only on `connections`? Or should `asyncConnectionPool` has a say in the decision?

**Review by ali-ince — APPROVED** [review]
LGTM! I added a couple of comments. Besides, having both `ConnectionPool` and `AsyncConnectionPool` in the code base started to be confusing for me, maybe it is time for a clean-up if those old pieces are not needed anymore.

**Comment by lutovich** [comment]
@ali-ince thanks for the review! I planned to get rid of those `*Async` suffixes and pool duality in the next step when doing blocking on top of async.

**Pull request merged**